New exact functions for exponential and logistic growth

jamesmaino opened the pull request:
Also a new simple module for founder effects.

jamesmaino committed:
fixed growth function with Euler solver, but have noticed this causes issues when numbers are small. Also ensured that pops arent growing during local dispersal

jamesmaino committed:
added founder effect module and exact growth functions

rafaqz commented:
looks great. Ill can merge when it passes the tests. Then you can update the scripts to match this commit and push

jamesmaino commented on src/growth.jl:
```diff
@@ -33,28 +50,54 @@ end
 
 # Rules
+# euler solver rules
+@inline rule(model::EulerExponentialGrowth, data, state, args...) = state + state * model.r
```
Just noticed this should be a `state + state * model.r * model.timestep`

rafaqz replied:
You can just keep pushing commits fixing things and theyll show up here

rafaqz replied:
But also I don't get the use of `model.timestep` here. Doesn't it means we have an increasing growth rate over time? Isn't exponential growth the outcome of running this multiple times as is?

jamesmaino replied:
I think ideally, we have a global timestep that is passed to Models(). The rate is dN/dt = r*N so the change in N with each timestep is dN = r*N*dT

rafaqz replied:
Yeah I got it after I sent that comment. timestep is timestep length, I was thinking it was t. 

We could add that to the Models and FrameData structs so it's set at the model level and passed to everything, as  cellsize is.

rafaqz commented on src/Dispersal.jl:
```diff
@@ -4,23 +4,24 @@ This package extends [Cellular.jl](https://github.com/rafaqz/Cellular.jl)
       Requires,
-      FieldMetadata
+      FieldMetadata,
+      Distributions
```
You'll need to add an entry for Distributions in the Requires file, this is breaking CI

jamesmaino replied:
So do you want me to do this? How does this usually proceed? n00b question :)

rafaqz replied:
Sorry I should have spelt it correctly with the caps! But yeah you should do it so you know for next time ;)

https://github.com/rafaqz/Dispersal.jl/blob/master/REQUIRE

Its a very simple file. All you have to do is add `Distributions` on a new line, so Pkg knows to download it. It probably doesn't need a version at this stage.

rafaqz commented on src/allee.jl:
```diff
@@ -0,0 +1,13 @@
+
+# Rules
+@inline rule(model::AlleeExtinction, data, state, args...) =
+    (state > model.minfounders ? state : 0)
```
This is perfect. It should work for float or int.

rafaqz commented on src/allee.jl:
```diff
@@ -0,0 +1,13 @@
+
+" Minimum individuals required for cell colonisation "
+@premix @columns struct MinimumFounders{MF}
+    minfounders::MF = 5 | true | (0.0, 100.0)
```
You could put this line directly in AlleeExtinction as its only used once. You'll also need to have `@columns` at the start to add the metadata, and MF in the brackets.

jamesmaino replied:
I have made the change (as well as others you commented on). But do I need to do another pull request?

rafaqz replied:
No just keep pushing to the same branch (master on jamesmaino/Dispersal.jl)  until the tests pass, then we can merge it

jamesmaino replied:
tests now pass and updates pushed.

rafaqz commented on src/growth.jl:
```diff
@@ -1,60 +1,103 @@
-@premix @columns struct GrowthRate{GR}
-    growthrate::GR = 1.1 | true | (0.0, 10.0)
+@premix @columns struct InstrinsicGrowthRate{GR, TS}
+    r::GR = 0.1 | true | (0.0, 10.0)
```
Maybe we should call this `rate` instead of `r`. Julia is kind of anti abbreviations like that except for short term vars. But maybe its ok too

jamesmaino replied:
Sure, I'm not too fussed. The standard notation for intrinsic growth rate is r.

jamesmaino committed:
minor error in growth module. also reduced redundancy in allee module.

jamesmaino committed:
updated intinsic to intrinsic typo
updated outward dispersal so that dispersal cannot increase pop size
fixed up growth and integration tests for growth modules and local 
dispersal

rafaqz committed:
Merge branch 'master' into master

rafaqz commented:
Merged! Awesome. Coveralls is grumpy we'll hae to add a few more tests.
